Refactor / Extend VideoOverlay Tutorial (ImageOverlay, SVGOverlay) (#8090)
Leaflet/Leaflet: docs/SlavaUkraini/examples/video-overlay/index.md

---
layout: tutorial_v2

> title: Overlay

Reviewer: Maybe use plural "Overlays"?

---

> # Overlay - Tutorial

Reviewer: This should be an h2 (`##`), other headings will need to be adjusted accordingly.

Author: got it, now I changed all headers for 1 level as it was before.
Before I tried to make it 1 level up since it was hard to distinguish #### (h4) from a simple paragraph

> ## Overlay - Tutorial

Reviewer: ```suggestion
## Overlays
```

Reviewer: same as above

There are 3 Overlays in Leaflet API: 
- [`ImageOverlay`](/reference.html#imageoverlay): Raster Layer, Extends [`Layer`](/reference.html#layer)
- [`VideoOverlay`](/reference.html#videooverlay): Raster Layer, Extends [`ImageOverlay`](/reference.html#imageoverlay)
- [`SVGOverlay`](/reference.html#svgoverlay): Vector Layer, Extends [`ImageOverlay`](/reference.html#imageoverlay)

In this tutorial, you’ll learn how to use these Overlays.

> ### `ImageOverlay`

Reviewer: Ideally this section should also have a live example, in a new `example-image.md`.

Author: I agree.
- image example added.

`L.imageOverlay` is used to load and display a single image over specific bounds of the map. 

To add an image overlay [`L.ImageOverlay`](/reference.html#imageoverlay) use this:
```

> 	var overlay = L.imageOverlay( imageUrl, bounds, options );

Reviewer: Most/all code blocks are indented too much.

Author: got it, now I changed all code to 0 level of indent for the most outer level like it is in all other tutorials

> `L.imageOverlay` is used to load and display a single image over specific bounds of the map. 

Reviewer 2: ```suggestion
`L.ImageOverlay` is used to load and display a single image over specific bounds of the map. 
```

```
First of all, create a Leaflet map and add a background `L.TileLayer` in the usual way:
```
var map = L.map('map').setView([37.8, -96], 4);

L.tileLayer('https://api.mapbox.com/styles/v1/{id}/tiles/{z}/{x}/{y}?access_token=' + mapboxAccessToken, {
	id: 'mapbox/satellite-v9',
	attribution: ...,
	tileSize: 512,
	zoomOffset: -1
}).addTo(map);
```
If you want to see the area which is covered by the ImageOverlay, you can add a [`L.Rectangle`](/reference.html#rectangle) with the same `L.LatLngBounds` to the map:
```
L.rectangle(latLngBounds).addTo(map);

map.fitBounds(latLngBounds);
```
Let's create an image overlay with multiple options:
```

> var imageUrl = 'http://www.lib.utexas.edu/maps/historical/newark_nj_1922.jpg',

Reviewer: ```suggestion
var imageUrl = 'https://www.lib.utexas.edu/maps/historical/newark_nj_1922.jpg',
```

var imageBounds = [[40.712216, -74.22655], [40.773941, -74.12544]];
var errorOverlayUrl = 'https://cdn-icons-png.flaticon.com/512/110/110686.png';

> var altText = 'Video of Hurricane Patricia from Satellite. Source: NASA'

Reviewer: alt text of image describes a video, should be updated.

> var bounds = L.latLngBounds([[ 32, -130], [ 13, -100]]);
> 
> var imageOverlay = L.imageOverlay( imageUrl, bounds, {

Reviewer 2: ```suggestion
var latLngBounds = L.latLngBounds([[ 32, -130], [ 13, -100]]);

var imageOverlay = L.imageOverlay( imageUrl, latLngBounds, {
```

	opacity: 0.8,
	errorOverlayUrl: errorOverlayUrl,
	alt: altText,
	interactive: true
}).addTo(map);
```
- `opacity` defines the opacity of the image overlay, it equals to `1.0` by default. Decrease this value to make an image overlay transparent and to expose the underlying map layer. 
	
- `errorOverlayUrl` is a URL to the overlay image to show in place of the overlay that failed to load.

- `alt` stands for the alt attribute of the image. Images alternatives (`alt` option) add valuable information for low vision or blind screen reader users. Image alternatives also benefit people who have poor or unstable internet, some cognitive disabilities. Moreover, it can improve the SEO of a website.

- `interactive` is `false` by default. If `true`, the image overlay will emit mouse events when clicked or hovered.

> You can find other options of `L.imageOverlay` in [`Docs`](/reference.html#imageoverlay)

Reviewer 2: ```suggestion
You can find other options of `L.ImageOverlay` in the [`docs`](/reference.html#imageoverlay)
```

---
layout: tutorial_v2
title: Overlays
---

## Overlays

There are 3 Overlays in Leaflet API: 
- [`ImageOverlay`](/reference.html#imageoverlay): Raster Layer, Extends [`Layer`](/reference.html#layer)
- [`VideoOverlay`](/reference.html#videooverlay): Raster Layer, Extends [`ImageOverlay`](/reference.html#imageoverlay)
- [`SVGOverlay`](/reference.html#svgoverlay): Vector Layer, Extends [`ImageOverlay`](/reference.html#imageoverlay)

In this tutorial, you’ll learn how to use these Overlays.

### `ImageOverlay`

`L.ImageOverlay` is used to load and display a single image over specific bounds of the map. 

To add an image overlay [`L.ImageOverlay`](/reference.html#imageoverlay) use this:
```
var overlay = L.imageOverlay( imageUrl, latLngBounds, options );
```
First of all, create a Leaflet map and add a background `L.TileLayer` in the usual way:
```
var map = L.map('map').setView([37.8, -96], 4);

L.tileLayer('https://api.mapbox.com/styles/v1/{id}/tiles/{z}/{x}/{y}?access_token=' + mapboxAccessToken, {
	maxZoom: 18,

> 	attribution: ...,

Reviewer: All other tutorials use a real attribution, e.g. `'Map data &copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors, Imagery © <a href="https://www.mapbox.com/">Mapbox</a>'`.

var imageOverlay = L.imageOverlay( imageUrl, latLngBounds, {
	opacity: 0.8,
	errorOverlayUrl: errorOverlayUrl,
	alt: altText,
	interactive: true
}).addTo(map);
```
If you want to see the area which is covered by the ImageOverlay, you can add a [`L.Rectangle`](/reference.html#rectangle) with the same `L.LatLngBounds` to the map:
```
L.rectangle(latLngBounds).addTo(map);
map.fitBounds(latLngBounds);
```
- `opacity` defines the opacity of the image overlay, it equals to `1.0` by default. Decrease this value to make an image overlay transparent and to expose the underlying map layer. 
	
- `errorOverlayUrl` is a URL to the overlay image to show in place of the overlay that failed to load.

- `alt` stands for the alt attribute of the image. Images alternatives (`alt` option) add valuable information for low vision or blind screen reader users. Image alternatives also benefit people who have poor or unstable internet, some cognitive disabilities. Moreover, it can improve the SEO of a website.

- `interactive` is `false` by default. If `true`, the image overlay will emit mouse events when clicked or hovered.

> You can find other options of `L.ImageOverlay` in the [`docs`](/reference.html#imageoverlay)

Reviewer: ```suggestion
You can find other options of `L.ImageOverlay` in the [documentation](/reference.html#imageoverlay).
```

	tileSize: 512,
	zoomOffset: -1
}).addTo(map);
```

Let's create an image overlay with multiple options:
```
var imageUrl = 'https://maps.lib.utexas.edu/maps/historical/newark_nj_1922.jpg';
var errorOverlayUrl = 'https://cdn-icons-png.flaticon.com/512/110/110686.png';
var altText = 'Image of Newark, N.J. in 1922. Source: The University of Texas at Austin, UT Libraries Map Collection.';
var latLngBounds = L.latLngBounds([[40.799311, -74.118464], [40.68202047785919, -74.33]]);

var imageOverlay = L.imageOverlay( imageUrl, latLngBounds, {
	opacity: 0.8,
	errorOverlayUrl: errorOverlayUrl,
	alt: altText,
	interactive: true
}).addTo(map);
```
If you want to see the area which is covered by the ImageOverlay, you can add a [`L.Rectangle`](/reference.html#rectangle) with the same `L.LatLngBounds` to the map:
```
L.rectangle(latLngBounds).addTo(map);
map.fitBounds(latLngBounds);
```
- `opacity` defines the opacity of the image overlay, it equals to `1.0` by default. Decrease this value to make an image overlay transparent and to expose the underlying map layer. 
	
- `errorOverlayUrl` is a URL to the overlay image to show in place of the overlay that failed to load.

> - `alt` stands for the alt attribute of the image. Images alternatives (`alt` option) add valuable information for low vision or blind screen reader users. Image alternatives also benefit people who have poor or unstable internet, some cognitive disabilities. Moreover, it can improve the SEO of a website.

Reviewer: ```suggestion
- `alt` sets the HTML [`alt`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-alt) attribute to provide an alternative text description of the image. Alternative text is essential information for screen reader users. It can also benefit people during poor network connectivity, in the case the image fails to load. Moreover, it can improve the SEO of a website.
```

A little rewrite to detail what `alt` really is and why it is useful. I also appreciate mentioning users with cognitive disabilities, though I think it is borderline TMI in this context, and perhaps better suited in https://leafletjs.com/SlavaUkraini/examples/accessibility/.

> title: Leaflet on Mobile

Reviewer: Side note: It seems this title has been wrong for quite some time, it should be updated.

Author: Yeah, I changed the name to _title: Overlay_, but not sure if it is good enough.

Author: Yeah, I changed it to the simple _title: Overlay_, but not sure if it is good enough to represent the tutorial.

title: Leaflet on Mobile
---

> ## VideoOverlay - Tutorial

Reviewer 2: Each Class (ImageOverlay, VideoOverlay, SVGOverlay) gets a own Headline with two `#`. Something like `## Adding a Image to the map`

Author: Thanks, now I understand what is a general idea of this refactoring.

	var overlay = L.imageOverlay( imageUrl, bounds, options );

`L.imageOverlay` is used to load and display a single image over specific bounds of the map. Extends [`Layer`](/reference.html#layer).

For a video overlay, just:

* Use `L.videoOverlay` instead of `L.imageOverlay`
* `L.videoOverlay` is used to load and display a video player over specific bounds of the map. Extends [`ImageOverlay`](/reference.html#imageoverlay). 
A video overlay uses the [`<video>`](https://developer.mozilla.org/docs/Web/HTML/Element/video) HTML5 element.
* Instead of the image URL, specify one video URL *or* an array of video URLs

```
	var videoUrls = [
		'https://www.mapbox.com/bites/00188/patricia_nasa.webm',
		'https://www.mapbox.com/bites/00188/patricia_nasa.mp4'
	];
        var errorOverlayUrl = 'https://cdn-icons-png.flaticon.com/512/110/110686.png';

>         var altText = 'Hurricane Patricia from Satellite. Source: NASA'

Reviewer 2: Please use tabs instead of spaces

	var bounds = L.latLngBounds([[ 32, -130], [ 13, -100]]);

	var videoOverlay = L.videoOverlay( videoUrls, bounds, {
		opacity: 0.8, 
                errorOverlayUrl: errorOverlayUrl,
                alt: altText,
                interactive: true,
                autoplay: true,
                muted: true
	}).addTo(map);
```

And just like that, you'll get the video on your map:

{% include frame.html url="example-nocontrols.html" %}

`errorOverlayUrl` is a URL to the overlay image to show in place of the overlay that failed to load. The option is inherited from [`ImageOverlay`](/reference.html#imageoverlay).

`alt` stands for the alt attribute of the image. Images alternatives (`alt` option) add valuable information for low vision or blind screen reader users. Image alternatives also benefit people who have poor or unstable internet, some cognitive disabilities. Moreover, it can improve the SEO of a website. The option is inherited from [`ImageOverlay`](/reference.html#imageoverlay).

> `interactive` is `false` by default. If `true`, the imageOverlay will emit mouse events when clicked or hovered. It means that no mouse events are triggered. The option is inherited from [`ImageOverlay`](/reference.html#imageoverlay).

Reviewer 2: Reorder the senteces, especially the last one.

`interactive: false`: No mouse events are triggered.
`interactive: true`: mouse events will be fired.

If a video can be shown in a webpage in this way, then Leaflet can display it inside a map. It is important that the videos are prepared in such a way that they will fit the map: The video should have a "north-up" orientation, and its proportions should fit the map. If not, it will look out of place.

> #### Creating a map with a background

Reviewer 2: ```suggestion
#### Creating a map
```

First of all, create a Leaflet map and add a background `L.TileLayer` in the usual way:
```
var map = L.map('map').setView([37.8, -96], 4);

L.tileLayer('https://api.mapbox.com/styles/v1/{id}/tiles/{z}/{x}/{y}?access_token=' + mapboxAccessToken, {
	id: 'mapbox/satellite-v9',
	attribution: ...,
	tileSize: 512,
	zoomOffset: -1
}).addTo(map);
```
Then, we'll define the geographical bounds that the video will cover. This is an instance of [`L.LatLngBounds`](/reference.html#latlngbounds), which is a rectangular shape:
```

> var bounds = L.latLngBounds([[ 32, -130], [ 13, -100]]);

Reviewer 2: ```suggestion
var latLngBounds = L.latLngBounds([[ 32, -130], [ 13, -100]]);
```

```
If you want to see the area covered by a `LatLngBounds`, use a [`L.Rectangle`](/reference.html#rectangle):
```

> L.rectangle(bounds).addTo(map);

Reviewer 2: ```suggestion
L.rectangle(latLngBounds).addTo(map);
```

> map.fitBounds(bounds);

Reviewer 2: ```suggestion
map.fitBounds(latLngBounds);
```

```
{% include frame.html url="example-bounds.html" %}

#### Adding the video overlay

Adding a video overlay works very similar to adding an image overlay. 

For a video overlay, just:

* Use `L.videoOverlay` instead of `L.imageOverlay`

> * `L.videoOverlay` is used to load and display a video player over specific bounds of the map. Extends [`ImageOverlay`](/reference.html#imageoverlay). 

Reviewer 2: ```suggestion
* `L.videoOverlay` is used to load and display a video player over specific bounds of the map. Extends [`L.ImageOverlay`](/reference.html#imageoverlay). 
```

A video overlay uses the [`<video>`](https://developer.mozilla.org/docs/Web/HTML/Element/video) HTML5 element.
* Instead of the image URL, specify one video URL *or* an array of video URLs

```
var videoUrls = [
	'https://www.mapbox.com/bites/00188/patricia_nasa.webm',
	'https://www.mapbox.com/bites/00188/patricia_nasa.mp4'
];
var errorOverlayUrl = 'https://cdn-icons-png.flaticon.com/512/110/110686.png';
var altText = 'Video of Hurricane Patricia from Satellite. Source: NASA'

> var bounds = L.latLngBounds([[ 32, -130], [ 13, -100]]);
> 
> var videoOverlay = L.videoOverlay( videoUrls, bounds, {

Reviewer 2: ```suggestion
var latLngBounds = L.latLngBounds([[ 32, -130], [ 13, -100]]);

var videoOverlay = L.videoOverlay( videoUrls, latLngBounds, {
```

	opacity: 0.8,
	errorOverlayUrl: errorOverlayUrl,
	alt: altText,
	interactive: true,
	autoplay: true,
	muted: true,
	playsInline: true

> }).addTo(map);

Reviewer: Update `examples/video-overlay/example.md` to reflect this?

Reviewer: @KonstantinBiryukov and also in `example-nocontrols.md`.

```
videoOverlay.getElement().pause();
```

This allows us to build custom interfaces. For example, we can build a small subclass of `L.Control` to play/pause this video overlay once it's loaded:

```
videoOverlay.on('load', function () {
	var MyPauseControl = L.Control.extend({
		onAdd: function() {
			var button = L.DomUtil.create('button');
			button.innerHTML = '⏸';
			L.DomEvent.on(button, 'click', function () {
				videoOverlay.getElement().pause();
			});
			return button;
		}
	});
	var MyPlayControl = L.Control.extend({
		onAdd: function() {
			var button = L.DomUtil.create('button');
			button.innerHTML = '▶️';
			L.DomEvent.on(button, 'click', function () {
				videoOverlay.getElement().play();
			});
			return button;
		}
	});

	var pauseControl = (new MyPauseControl()).addTo(map);
	var playControl = (new MyPlayControl()).addTo(map);
});
```

{% include frame.html url="example.html" %}

### `SVGOverlay`
`L.imageOverlay` is used to load and display a single image over specific bounds of the map. 

To add an image overlay [`L.ImageOverlay`](/reference.html#imageoverlay) use this:
```
var svgOverlay = L.svgOverlay( SVGElement, bounds, options );

> ```

Reviewer 2: Copy & Past issue?

Author: yeah, probably. fixed

It instantiates an image overlay object given an SVG element and the geographical bounds it is tied to. A viewBox attribute is required on the SVG element to zoom in and out properly.

#### `Creating SVG element`
Let's create SVG element:
```
var svgElement = document.createElementNS("http://www.w3.org/2000/svg", "svg");
svgElement.setAttribute('xmlns', "http://www.w3.org/2000/svg");
svgElement.setAttribute('viewBox', "0 0 200 200");
svgElement.innerHTML = '<rect width="200" height="200"/><rect x="75" y="23" width="50" height="50" style="fill:red"/><rect x="75" y="123" width="50" height="50" style="fill:#0013ff"/>';
```

Alternatively, you can create svg element in html code:
```
<svg id="image" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 200 200"><rect width="200" height="200"/><rect x="75" y="23" width="50" height="50" style="fill:red"/><rect x="75" y="123" width="50" height="50" style="fill:#0013ff"/></svg>
```

And choose this SVG element by using a querySelector:

```
var svgElement = document.querySelector('#image');
```

#### `Creating SVG Overlay`
Create SVGOverlay with desired options similarly to ImageOverlay and VideoOverlay:
```

> var bounds = L.latLngBounds([[ 32, -130], [ 13, -100]]);
> map.fitBounds(bounds);

Reviewer 2: ```suggestion
var latLngBounds = L.latLngBounds([[ 32, -130], [ 13, -100]]);
map.fitBounds(latLngBounds);
```

> var svgOverlay = L.svgOverlay(svgElement, bounds, {

Reviewer 2: ```suggestion
var svgOverlay = L.svgOverlay(svgElement, latLngBounds, {
```

	opacity: 0.7,
	interactive: true
}).addTo(map);
```

> Although SVGOverlay does not have its own unique options, it inherited a variety of options from ImageOverlay, Interactive layer and Layer.

Reviewer: "it inherited" is past tense, shouldn't it be "it inherits"?

> Check out Documentation to find out more [`SVGOverlay`](/reference.html#svgoverlay) options.

Reviewer 2: ```suggestion
Check out Documentation to find out more [`L.SVGOverlay`](/reference.html#svgoverlay) options.
```

If a video can be shown in a webpage in this way, then Leaflet can display it inside a map. It is important that the videos are prepared in such a way that they will fit the map: The video should have a "north-up" orientation, and its proportions should fit the map. If not, it will look out of place.

#### Creating a map

First of all, create a Leaflet map and add a background `L.TileLayer` in the usual way:
```
var map = L.map('map').setView([37.8, -96], 4);

L.tileLayer('https://api.mapbox.com/styles/v1/{id}/tiles/{z}/{x}/{y}?access_token=' + mapboxAccessToken, {
	id: 'mapbox/satellite-v9',

> 	attribution: ...,

Reviewer: Use real attribution here too.

	'https://www.mapbox.com/bites/00188/patricia_nasa.webm',
	'https://www.mapbox.com/bites/00188/patricia_nasa.mp4'
];
var errorOverlayUrl = 'https://cdn-icons-png.flaticon.com/512/110/110686.png';
var altText = 'Video of Hurricane Patricia from Satellite. Source: NASA'
var latLngBounds = L.latLngBounds([[ 32, -130], [ 13, -100]]);

var videoOverlay = L.videoOverlay( videoUrls, latLngBounds, {
	opacity: 0.8,
	errorOverlayUrl: errorOverlayUrl,
	alt: altText,
	interactive: true,
	autoplay: true,
	muted: true,
	playsInline: true
}).addTo(map);
```

And just like that, you'll get the video on your map:

{% include frame.html url="example-nocontrols.html" %}

- `autoplay` option defines whether the video starts playing automatically when loaded. It is `true` by default. It is important to know that on some browsers `autoplay` functionality will only work with `muted` option explicitly set to `true`.

- `muted` option defines whether the video starts on mute when loaded. It is `false` by default.

- `playsInline` option when it is set to `true` allows video to play inline without automatically entering fullscreen mode when playback begins in the mobile browser. It is `true` by default.

> You can find other options of `L.videoOverlay` in [`Docs`](/reference.html#videooverlay)

Reviewer: ```suggestion
You can find other options of `L.videoOverlay` in the [documentation](/reference.html#videooverlay).
```

Spell out "documentation", for consistency with other tutorials.

L.tileLayer('https://api.mapbox.com/styles/v1/{id}/tiles/{z}/{x}/{y}?access_token=' + mapboxAccessToken, {
	id: 'mapbox/satellite-v9',
	attribution: 'Map data &copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors, ' +
		'Imagery © <a href="https://www.mapbox.com/">Mapbox</a>',
	tileSize: 512,
	zoomOffset: -1
}).addTo(map);
```
Then, we'll define the geographical bounds that the video will cover. This is an instance of [`L.LatLngBounds`](/reference.html#latlngbounds), which is a rectangular shape:
```
var latLngBounds = L.latLngBounds([[ 32, -130], [ 13, -100]]);
```
If you want to see the area covered by a `LatLngBounds`, use a [`L.Rectangle`](/reference.html#rectangle):
```
L.rectangle(latLngBounds).addTo(map);

map.fitBounds(latLngBounds);
```
{% include frame.html url="example-bounds.html" %}

#### Adding the video overlay

Adding a video overlay works very similar to adding an image overlay. 

For a video overlay, just:

* Use `L.videoOverlay` instead of `L.imageOverlay`
* `L.videoOverlay` is used to load and display a video player over specific bounds of the map. Extends [`L.ImageOverlay`](/reference.html#imageoverlay). 

> A video overlay uses the [`<video>`](https://developer.mozilla.org/docs/Web/HTML/Element/video) HTML5 element.

Reviewer: ```suggestion
A video overlay uses the [`<video>`](https://developer.mozilla.org/docs/Web/HTML/Element/video) HTML element.
```

```
videoOverlay.on('load', function () {
	var MyPauseControl = L.Control.extend({
		onAdd: function() {
			var button = L.DomUtil.create('button');
			button.innerHTML = '⏸';
			L.DomEvent.on(button, 'click', function () {
				videoOverlay.getElement().pause();
			});
			return button;
		}
	});
	var MyPlayControl = L.Control.extend({
		onAdd: function() {
			var button = L.DomUtil.create('button');
			button.innerHTML = '▶️';
			L.DomEvent.on(button, 'click', function () {
				videoOverlay.getElement().play();
			});
			return button;
		}
	});

	var pauseControl = (new MyPauseControl()).addTo(map);
	var playControl = (new MyPlayControl()).addTo(map);
});
```

{% include frame.html url="example-video.html" %}

### `SVGOverlay`
`L.SVGOverlay` is used to load, display and provide DOM access to an SVG file over specific bounds of the map. 

To add an SVG overlay [`L.SVGOverlay`](/reference.html#svgoverlay) use this:
```
var svgOverlay = L.svgOverlay( SVGElement, svgElementBounds, options );
```
It instantiates an image overlay object given an SVG element and the geographical bounds it is tied to. A viewBox attribute is required on the SVG element to zoom in and out properly.

> #### `Creating SVG element`

Reviewer: ```suggestion
#### `Creating an SVG element`
```

> Let's create SVG element:

Reviewer: ```suggestion
Let's create an SVG element:
```

```
var svgElement = document.createElementNS("http://www.w3.org/2000/svg", "svg");
svgElement.setAttribute('xmlns', "http://www.w3.org/2000/svg");
svgElement.setAttribute('viewBox', "0 0 200 200");
svgElement.innerHTML = '<rect width="200" height="200"/><rect x="75" y="23" width="50" height="50" style="fill:red"/><rect x="75" y="123" width="50" height="50" style="fill:#0013ff"/>';
```

> Alternatively, you can create svg element in html code:

Reviewer: ```suggestion
Alternatively, you can create the SVG element in HTML code:
```

Adding a video overlay works very similar to adding an image overlay. 

For a video overlay, just:

* Use `L.videoOverlay` instead of `L.imageOverlay`
* `L.videoOverlay` is used to load and display a video player over specific bounds of the map. Extends [`ImageOverlay`](/reference.html#imageoverlay). 
A video overlay uses the [`<video>`](https://developer.mozilla.org/docs/Web/HTML/Element/video) HTML5 element.
* Instead of the image URL, specify one video URL *or* an array of video URLs

```
	var videoUrls = [
		'https://www.mapbox.com/bites/00188/patricia_nasa.webm',
		'https://www.mapbox.com/bites/00188/patricia_nasa.mp4'
	];
	var errorOverlayUrl = 'https://cdn-icons-png.flaticon.com/512/110/110686.png';
	var altText = 'Video of Hurricane Patricia from Satellite. Source: NASA'
	var bounds = L.latLngBounds([[ 32, -130], [ 13, -100]]);

	var videoOverlay = L.videoOverlay( videoUrls, bounds, {
		opacity: 0.8,
		errorOverlayUrl: errorOverlayUrl,
		alt: altText,
		interactive: true,
		autoplay: true,
		muted: true,
		playsInline: true
	}).addTo(map);
```

And just like that, you'll get the video on your map:

{% include frame.html url="example-nocontrols.html" %}

> - `autoplay` option defines whether the video starts playing automatically when loaded. It is `true` by default. It is important to know that on some browsers `autoplay` functionality will only work with `muted` option explicitly set to `true`. It is important to know that on some browsers `autoplay` functionality will only work with `muted` option explicitly set to `true`.

Reviewer: > It is important to know that on some browsers `autoplay` functionality will only work with `muted` option explicitly set to `true`.

is stated twice here.

```

{% include frame.html url="example.html" %}

## `SVGOverlay`
`L.imageOverlay` is used to load and display a single image over specific bounds of the map. 

To add an image overlay [`L.ImageOverlay`](/reference.html#imageoverlay) use this:
```
	var svgOverlay = L.svgOverlay( SVGElement, bounds, options );
```
It instantiates an image overlay object given an SVG element and the geographical bounds it is tied to. A viewBox attribute is required on the SVG element to zoom in and out properly.

### 1. Creating SVG element

> ##### Approach #1: 

Reviewer: Is this heading necessary?

Author: removed, thanks

Let's create SVG element:
```
	var svgElement = document.createElementNS("http://www.w3.org/2000/svg", "svg");
	svgElement.setAttribute('xmlns', "http://www.w3.org/2000/svg");
	svgElement.setAttribute('viewBox', "0 0 200 200");
	svgElement.innerHTML = '<rect width="200" height="200"/><rect x="75" y="23" width="50" height="50" style="fill:red"/><rect x="75" y="123" width="50" height="50" style="fill:#0013ff"/>';
```

> ##### Approach #2: 

Reviewer: Is this heading necessary?
